Reduce chances of assistant using patchCard when user wants to edit gts file

jurgenwerk opened the pull request:
This situation was getting very problematic, especially in Claude 3.7 Sonnet:

<img width="477" alt="image" src="https://github.com/user-attachments/assets/8ad4ad9c-bf0b-41b8-9c3a-41d0d6d433a1" />

<img width="553" alt="image" src="https://github.com/user-attachments/assets/089b22e6-de7d-4f36-b834-dbb6842a2668" />

jurgenwerk commented on packages/ai-bot/helpers.ts:
```diff
@@ -681,14 +681,24 @@ Attached files:
+  );
+
+  if (cardPatchTool && sourceCodeEditingSkillCard) {
+    systemMessage += `First determine if the user's request involves template editing, visual changes, or styling, or card logic changes.
```
Based on my experiments it's far more reliable if we put this message at the beginning, before skill prompts are added.

lukemelia replied:
I don't really like the idea of special casing some skills in the aibot.

jurgenwerk replied:
Right, as discussed I think renaming the ambiguous `patchCard` could be beneficial in a way where we won't have to include special cases

jurgenwerk commented on packages/base/SkillCard/boxel-coding.json:
```diff
@@ -2,7 +2,7 @@
     "type": "card",
     "attributes": {
-      "instructions": "Boxel is a platform where people can create Cards, which under the hood are built out of glimmer components and ember.\n\nCards are independent linkable items that get an ID. Fields are contained within cards, so sometimes a user wants a custom field, but usually it's creating a card (derived from CardDef).\n\nUse glimmer templating and typescript for the code. Remember the ...
+      "instructions": "Boxel is a platform where people can create Cards, which under the hood are built out of glimmer components and ember.\n\nCards are independent linkable items that get an ID. Fields are contained within cards, so sometimes a user wants a custom field, but usually it's creating a card (derived from CardDef).\n\nUse glimmer templating and typescript for the code. Remember the ...
```
Added: `<style> tag must be at the root of the <template>, otherwise the app will break. It must not be nested.`. I got this error after code patch so I thought it would be a good idea to add this instruction

jurgenwerk commented on packages/base/SkillCard/card-editing.json:
```diff
@@ -2,7 +2,7 @@
     "type": "card",
     "attributes": {
-      "instructions": "- If the user wants the data they see edited, AND the patchCard function is available, you MUST use the \"patchCard\" function to make the change.\n- If the user wants the data they see edited, AND the patchCard function is NOT available, you MUST ask the user to open the card and share it with you.\n- If you do not call patchCard, the user will not see the change.\n- You c ...
+      "instructions": "- If the user wants the card data they see edited, AND the patchCard function is available, you MUST use the \"patchCard\" function to make the change.\n- If the user wants the data they see edited, AND the patchCard function is NOT available, you MUST ask the user to open the card and share it with you.\n- If you do not call patchCard, the user will not see the change.\n-  ...
```
Added: `- Do NOT use patchCard function when user is asking for template changes, visual modifications, style updates, or card behavior changes. patchCard should ONLY be used when explicitly changing card data values bound to the card definition in gts files, or data relationships, and nothing else."`

jurgenwerk commented on packages/base/SkillCard/source-code-editing.json:
```diff
@@ -2,7 +2,7 @@
     "type": "card",
     "attributes": {
-      "instructions": "To alter code for the user, you must use a SEARCH/REPLACE block.\n\nA SEARCH/REPLACE block has three sections. A file url for the file to update, a section of code to search for, and the code to replace it with. All code within the SEARCH will be replaced.\n\nThis is ABSOLUTELY CRUCIAL, WITHOUT THIS THE CODE PATCH WON'T WORK: in the beginning of the code block, before <<<<< ...
+      "instructions": "To alter code for the user, you must use a SEARCH/REPLACE block.\n\nA SEARCH/REPLACE block has three sections. A file url for the file to update, a section of code to search for, and the code to replace it with. All code within the SEARCH will be replaced.\n\nThis is ABSOLUTELY CRUCIAL, WITHOUT THIS THE CODE PATCH WON'T WORK: in the beginning of the code block, before <<<<< ...
```
Added: `if you recognize the user wants to edit a template, do a visual change to a card, or describe a certain style, then you must use a SEARCH/REPLACE block to perform an edit to the attached gts file, by default in the isolated template. Do not default to using the patchCard tool function, unless the user asks you to change the supporting data of the card (bound model values for fields).\n\nWhen you respond with a SEARCH/REPLACE block, do not refer to it as a SEARCH/REPLACE block, as this is an internal structure that will get shown to user in a different format. If you need to refer to it, talk about it in a semantic way. For example, do not say 'I'll use a SEARCH/REPLACE block to add a template', but rather 'I'll add a border around the section', for example`

jurgenwerk commented:
I think we wouldn't need to bother explaining this so much if we renamed `patchCard` to something more exact. Right now it's confusing and ambiguous because it's not clear what exactly it's patching.

habdelra commented:
> I think we wouldn't need to bother explaining this so much if we renamed `patchCard` to something more exact. Right now it's confusing and ambiguous because it's not clear what exactly it's patching.

Agreed. perhaps patch instance. and maybe we can discuss in the skill that a card encapsulates behavior and visual design which is implemented as a class, and an instance is an instantiation of a card class that is populated with data. And then we rename `patchCard` to `patchInstance`. In general even in our own code it can be super confusing about whether we are talking about instances or card classes as we use the word "card" so interchangeably. I think AI is just picking up on that confusion. This is actulaly what inspired me to clean up the terminology in the store service as it started to get really hard to reason about.

lukemelia commented:
Maybe `UpdateCardInstance`

jurgenwerk commented:
Ok I'll try to rename to reduce confusion.

jurgenwerk committed:
patchCard -> patchCardInstance (to remove confusion in the AI when working with card defs and instances)

jurgenwerk committed:
Adjust skills to improve AI cognition on when to use card instance patching vs card definition patching

jurgenwerk committed:
Fix skill description in tests

burieberry reviewed (approved):
The name change looks good to me

jurgenwerk committed:
When checking skill diff, consider only the skill jsons (not all jsons)

jurgenwerk commented:
This is now ready for another review. Key things:

- Compared to my original PR, I removed extra instructions added to the system prompt. Extra instructions are now added in skills only
- `patchCard` was removed to `patchCardInstance`
- Added some slight adjustments to skills instructions

The result of that is that I am now unable to reproduce this type of problems where the assistant will fail to understand you want to change the card definition (and will use patchCard instead, which doesn't make too much sense):
<img width="330" alt="image" src="https://github.com/user-attachments/assets/eada4541-52ea-4d6f-a741-a745f59f8a7c" />

<img width="800" alt="image" src="https://github.com/user-attachments/assets/9a892e88-6383-4da7-82f0-d72ea12aa070" />

jurgenwerk commented on .github/workflows/diff-skills.yaml:
```diff
@@ -30,8 +30,8 @@ jobs:
-          # Get the list of changed JSON files
-          CHANGED_FILES=$(echo "${{ steps.changed-files.outputs.all_changed_files }}" | tr ' ' '\n')
+          # Get the list of changed JSON files (with a "Skill" in the file path)
+          CHANGED_FILES=$(echo "${{ steps.changed-files.outputs.all_changed_files }}" | tr ' ' '\n' | grep -E 'Skill')
```
There are several other json files in our repo, we aren't interested in those here, and their content isn't what diff-skills expects it to be, leading to parsing errors
